src/main/java/hudson/remoting/Util.java

```diff
@@ -131,4 +136,37 @@ static URLConnection openURLConnection(URL url) throws IOException {
         }
         return con;
     }
+    
+    static InetSocketAddress getResolvedHttpProxyAddress(String host, int port) throws IOException {
+    	InetSocketAddress targetAddress = null;
+    	Iterator<Proxy> proxies = ProxySelector.getDefault().select(URI.create(String.format("http://%s:%d", host, port))).iterator();
+    	// No break -> The while condition already checks if targetAddress is null.
+    	while (targetAddress == null && proxies.hasNext()) {
+    		Proxy proxy = proxies.next();
+    		if (proxy.type() == Proxy.Type.HTTP) {
+    			if (!(proxy.address() instanceof InetSocketAddress))
+    				throw new IOException("Unsupported proxy address type " + proxy.address().getClass());
```

Comment:
<code>proxy.address</code> may return null if it is a direct connection. In such case there will be an NPE in the routine. 
If I understand the code correctly, <code> getResolvedHttpProxyAddress</code> should return null in such case to invoke the no-proxy logic in the calling method